### DISCUSSION TITLE
.NET 8 / Blazor / Oqtane 5

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#3448] [General]
As I have gotten further into Blazor in .NET 8 I have come to realize that these are not just some new Blazor features - they are in fact a completely new application development model which is extremely powerful... but are also going to be highly disruptive to the .NET developer community. 

Since Blazor was first introduced in 2018, it's capabilities have closely aligned to its tagline "Blazor is a feature of ASP.NET for building interactive web UIs using C# instead of JavaScript". Specifically, the word "interactive" needs to be emphasized because the entire development model was based on the notion of interactivity. Essentially you can think of Blazor as providing a container for building interactive applications - and those containers could be hosted in different environments (ie. server, browser, desktop) by using different technology approaches (ie. SignalR, WebAssembly, WebView2). 

Interactivity has always been a core concept of Blazor... a default behavior... and therefore it is also a default dependency which all Blazor components rely upon to function correctly. Please note that I am not only talking about custom Blazor applications... I am also talking about all existing Blazor UI component libraries, etc.... the entire Blazor ecosystem today is built on a foundation of interactivity.

So what is "interactivity"?

Any interaction by a user which occurs client-side can be considered interactivity. In its simplest form you can think of a button with an onclick event and handler, or a dropdown list with an onchange event and handler - these require interactivity. So with that context in mind, it should be very obvious just how pervasive this concept is in existing Blazor applications (regardless of hosting model) and also how dependent these applications are on interactivity.

This all changes with .NET 8....

Under the new model, Blazor is NOT interactive by default. It is server-generated and non-interactive. So instead of viewing Blazor as a container for building interactive applications, Microsoft is expecting developers to completely change their mindset and perspective related to Blazor. 

Going forward Blazor is a non-interactive container for building applications... which also includes the capability for individual components to be interactive. Interactivity becomes a "feature" rather than a default behavior/expectation. 

This is a very significant change. 

It means that during development you can no longer assume that everything is interactive. You will need to invest some time and energy into dissecting your application into interactive vs non-interactive scenarios. And the challenge is that some of the development approaches which you have relied upon in the past in the interactive Blazor model, are not compatible with the new non-interactive model. So in many cases it is not simply a matter of adding a "non-interactive" attribute to a component - it will require refactoring the component code. So you will need to be diligent in how you approach your Blazor development efforts. 

You will also want to choose your dependencies wisely... for example if you are using a specific UI component library throughout your application which only supports interactive scenarios, it means your application will not be able to get the benefits of the non-interactive approach.

On the positive side, this new approach opens up a whole new world of opportunities. It creates the potential for much more scalable applications which have faster startup time, and consume fewer resources. It creates opportunities for new light-weight UI components which do not have to depend on SignalR or WebAssembly for interactivity. It allows for the use of standard JavaScript without jumping through JS Interop hoops. And it allows interactive capabilities to be sprinkled in as needed.

What is the reason for this change?

Interactivity is expensive. It requires more infrastructure, processing, and resources. Server-side rendering is much more efficient and since the majority of content in web applications is static anyways, it makes sense that it should be the default rendering mode. And although Microsoft has never officially said this, it does make sense for them to consolidate their various approaches to web development so that they can focus on fewer options in the future. 

Is this new non-interactive Blazor approach required in .NET 8?

No. You can upgrade an existing Blazor application by simply updating the target framework and packages, and it will function just fine. The legacy Blazor Server and Blazor WebAssembly fully interactive hosting models are still supported. However, I am fairly certain that Microsoft will recommend that developers migrate their apps to the new non-interactive approach as it is the future of Blazor. 

What does this mean for Oqtane?

Blazor in .NET 8 is very much aligned with Oqtane's vision. Specifically, the concept of composing an application dynamically from a variety of modular components (ie. "islands of interactivity") rather than focusing on a tightly coupled monolithic application, is core to Oqtane's philosophy. The fact that these modular components will be able to embrace both interactive and non-interactive approaches is a huge benefit for Oqtane developers. And since the fundamental building blocks have remained the same ie. razor components, razor class libraries, etc... the existing core framework capabilities can be preserved and the migration path for Oqtane developers will not be challenging (existing modules should continue to function just fine in .NET 8 - they will default to interactive rendering for backward compatibility).

It also means that some of my earlier assumptions of how Oqtane could integrate with .NET 8 were incorrect. It looks like Microsoft provided enough extensibility that Oqtane can opt out of certain behaviors and still get the benefits of the new Blazor approach. There is still quite a bit of work to do but I am hopeful it will come together soon. It is very likely that I will not be reviewing/merging other types of PRs in the coming weeks as I focus on getting Oqtane 5.0 across the finish line.

**Comment by W6HBR** [comment]
@sbwalker  I appreciate you taking the time to digest and explore some of the nuances of .NET 8 and boiling it down to concise information for how it will (or will not) affect us as Oqtane implementors.

Kudos and thank you for keeping us informed.

**Comment by sbwalker** [comment]
I am going to include some detailed notes of the migration challenges in this thread. From a high level, the guidance from Microsoft is summarized below... however it is important to understand that these migration steps explain how to migrate a Blazor Server application created from the default Blazor template to .NET 8 - so basically a "Hello World" application that has no actual functionality. Migrating a real world Blazor application like Oqtane is going to be much more complicated.

https://github.com/danroth27/Net8BlazorServer/blob/b880f6e7a175b483918f6c77a754e7a99b46f4dc/README.md

1. Update the project file to target net8.0 and package references to the latest .NET 8 versions.

2. Move the contents of App.razor to Routes.razor.

3. Move content from Pages/_Host.cshtml to App.razor:
- Update script to blazor.web.js.
- Replace tag helpers with Blazor equivalents.

4. Update Program.cs:
- Replace AddServerSideBlazor() with AddRazorComponents().AddInteractiveServerComponents().
- Add app.UseAntiforgery().
- Replace MapBlazorHub() with MapRazorComponents<App>().AddInteractiveServerRenderMode().
- Remove app.MapFallbackToPage("/_Host");

**Comment by sbwalker** [comment]
1. Update the project file to target net8.0 and package references to the latest .NET 8 versions.

This is fairly strightforward - just tedious. The changeset is here: https://github.com/oqtane/oqtane.framework/commit/9f28ee2982fb7669a39c83eb74e6451992a304d8. Note that when .NET ships its final version, all of the package references will need to be updated with production versions.

**Comment by sbwalker** [comment]
2. Move the contents of App.razor to Routes.razor.

In the default Blazor template in .NET 8 you will notice that the Routes.razor component is in the Server project. In Oqtane the App.razor is in the Client project. Attempting to move the App.razor to the Server project revealed all of the Client dependencies it relies upon. I do not understand why this component needs to be in the Server project so I decided to leave it in the Client project for now. So I simply renamed to Routes.razor.

**Reply by thabaum** [reply]
I had tried this all in Oqtane already I had to stop along the way to deal with some other things.  I was having some issues I believe if I recall and wanted to try today.  I was not sure about the moving of the file to the server project which maybe necessary I am sensing.  The custom router component is what I was trying to get the project to allow auto mode for interactivity as my first goal and what I am most interested in for the entire project, however per component with SSR was my other goal to see the two at work and know I am actually making some kind of progress towards making these new .NET 8 features function.  This is where things get really interesting in how to deal with making Oqtane fully featured in .NET 8.  I pretty much stopped at this point to wait for more input on how things would go so I didn't go too far off coarse.

**Reply by sbwalker** [reply]
@thabaum "per component with SSR" is the exact same thing as "auto mode" - they both rely on the same concepts.

**Reply by thabaum** [reply]
@sbwalker To test I was thinking it would be a per-component for downloading the WASM .dll to the client only if the component is set to download it.  I was under the impression that with SSR the client would not download the WebAssembly dll package unless instructed to, where with "Auto" mode it will download the DLL for WebAssembly right after first render no matter what. I suppose the options for "Global" and "Per Component" along with the different render modes is where developers need to pay attention when making components.  Trying to fully understand all the differences and how results could vary.

For example with "Auto" mode I was under the impression that it will download the DLL no matter what, where SSR with Per Component enabled would only do this if a component needed allowed it and ONLY do it for that page, not every or any page.  As in if you go to a new page the WebAssembly .dll would be dismissed and the client browser would use static server rendering with no interactivity.  With Auto mode and "global" settings you get the WebAssembly DLL for the entire application regardless, first downloading the server side rendering of the page, then the client package is loaded into the browser... if that makes sense?

**Reply by sbwalker** [reply]
@thabaum actually this is not correct - Oqtane will still need to take responsibility for downloading client module/theme assemblies itself when using WebAssembly for interactivity. Blazor does not support this level of modularity and .NET 8 does not include any additional capability in this area.

**Comment by sbwalker** [comment]
3. Move content from Pages/_Host.cshtml to App.razor:

There are a LOT of changes required in this area as the _Host.cshtml razor page contains a variety of functionality which does not migrate directly to a razor component.

**Comment by sbwalker** [comment]
4. Update Program.cs:

The standard configuration is not going to work for Oqtane. This is because MapRazorComponents() is designed for Blazor applications which use routable components (ie. @page) and the default Blazor router. Oqtane uses standard components and a custom router. So there will need to be a custom routing integration created which allows routes to be mapped to the Oqtane root component. Microsoft provided a dynamic routing example which may be useful: https://github.com/javiercn/DynamicRoutingBlazor/blob/main/DynamicRoutingBlazor/DynamicRouting/CustomResolver.cs

**Comment by thabaum** [comment]
@sbwalker With all the major changes with updates to Oqtane relating to .NET 8 would it be worth it to also move the `Startup.cs` into the `Program.cs` or is it best to have the 2 files?

**Reply by sbwalker** [reply]
@thabaum I am trying to focus on only the Blazor SSR changes at this point - any other change could introduce new issues which make the Blazor SSR changes even more difficult to integrate.

**Reply by sbwalker** [reply]
@thabaum I think the challenge you are having is including MapRazorComponents() in Startup.cs (since the examples all use Program.cs instead). You can simply do the following:

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapRazorComponents<App>()
                    .AddInteractiveServerRenderMode()
                    .AddInteractiveWebAssemblyRenderMode();
            }); 

However this is not going to get you anywhere because Oqtane does not use routable components.

**Reply by thabaum** [reply]
@sbwalker To be clear on the different types of rendering configurations in .NET 8 from what I can recall and have tested, there are 3 types with 4 options more or less which can be set in the application as "Global" or "Per Component Per Page":

Static Server Rendering (SR*) = Static Server Rendering ( NO Interactivity)
Interactive Server (SSR)= Server-Side Rendering with interactivity (SignalR)
Interactive WebAssembly (IWA*)= Client-Side Rendering with interactivity
Interactive Auto (IA*)=Client-Side Rendering using Server-Side Rendering initially then WebAssembly on Subsequent visits

* not sure what to say for short... since "Static Server Rendering" and "Server-Side Rendering" use the same characters it is a bit confusing.  But in all I don't think Oqtane uses much "Static Server Rendering" and NodsJS calls it SSR and SSG so maybe that is where I need to clear things up for myself.  Static Server Rendering is probably like Site Site Generation or Static Site which builds all at runtime.  SSR is server-side so working on understanding "NodeJS" sure helps understand what Blazor is trying to do.

[https://nextjs.org/docs/pages/building-your-application/rendering](https://nextjs.org/docs/pages/building-your-application/rendering)

[https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0#render-modes](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0#render-modes)

After working with all the examples for creating the different types and seeing them work first hand from the examples they provide to test this functionality out with the new Blazor Web Application template.  I created the project and made a page for each of the 11 render examples I can post up in GitHub I suppose for others to test.

These this section of the page shows the 11 examples (one page per component example) plus a few after I have not tested for making an entire application use the different render modes.

[https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0#static-render-mode](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0#static-render-mode)

This was last week so I am trying to chase back my thoughts as I tried to hack Oqtane to get one of the render modes to work as expected.  Thing was I had no real way to test it. 

While thinking about it I am feeling the need to create modules that perform each type of feature similar to the render mode examples project I created to see them work and not work as expected.

These examples here also to potentially make some components to test from:
[https://github.com/dotnet/blazor-samples/tree/main/8.0/BlazorWebAppEFCore/Components/Pages](https://github.com/dotnet/blazor-samples/tree/main/8.0/BlazorWebAppEFCore/Components/Pages)

I think it would be good to have some basic example pages with components using each of the render mode examples to test fully different render mode behaviors to be expected.  So I am going to focus on creating these properly next hoping I can see some expected results manually changing settings within the Oqtane Framework's settings for now.

Fairly busy until middle of the week but trying to squeeze in what I can to prepare for an event of solid non-stop testing and investigating more without any interruptions.  

I appreciate you sharing what you have so far related Oqtane and .NET 8.  I want to be sure I can help test out what I can when solutions are found.

One thing I am wanting to do is change the render mode from @render-mode to @rendermode for example in the component related to  the Oqtane router.  Another thing is adding a way to set the way a module is rendered from Module Settings possibly.

I will wait and look forward to hearing more as you spend some time digging in.

Thank you.

**Comment by sbwalker** [comment]
I was able to create a minimal POC for Oqtane SSR today using the Blazor Web template (with WebAssembly chosen as the Interactive option). 

-	Does not use any @page components (all @page components migrated to Client project and converted into standard components)
-	Uses a fallback route mapped to the App root component
-	Uses a custom SiteRouter rather than the default Blazor router (which simply maps the current Url path to some Module metadata)
-	The SiteRouter is implemented in the Server project and uses Server rendering 
-	The SiteRouter has a variety of child components to simulate the nested behavior of components in Oqtane (ie. themes -> containers -> modules)
-	Uses DynamicComponent and RenderFragment to construct the UI dynamically

The project compiles and runs… however there are a variety of scenarios which do not behave as expected. I have reported the issues to the .NET team and I am waiting on a response.

**Reply by iJungleboy** [reply]
@sbwalker thanks so much for your work!

I couldn't attend the meeting yesterday but I personally believe that there are a lot of scenarios where high-performance, pure HTML output is super important - especially in all our CMS scenarios as well as Content of applications. And IMHO all applications nowadays have content.

This also includes SEO things. For example, oqtane.org ATM doesn't work for PageSpeed (no preview etc.) 
https://pagespeed.web.dev/analysis/https-www-oqtane-org/c8tm29l6fe?form_factor=mobile

We got it to work for blazor-cms.org but we needed some magic for it to work:
https://pagespeed.web.dev/analysis/https-blazor-cms-org/m0o242ibmi?form_factor=mobile

So these kinds of improvements will certainly help a lot - thanks!

**Reply by sbwalker** [reply]
@iJungleboy I agree that the new capabilities in Blazor have a lot of value. Unfortunately the migration effort is proving to be challenging for Oqtane... and I believe it will be challenging for anyone attempting to migrate an existing Blazor application. One of the challenges is the lack of documentation/examples for real world scenarios - the only guidance currently is for Hello World apps which is not very useful. I expect this situation to improve once .NET 8 is officially released.

**Comment by sbwalker** [comment]
Please refer to this new repo for a POC related to Blazor SSR on .NET 8:  https://github.com/oqtane/OqtaneSSR
